api: use contexts

Kulezi opened the pull request:
This PR changes NewSession, Session.Prepare, Query.Execute, Query.AsyncExecute to take context.Context as an argument.
Context passed to NewSession forces it to close on context cancellation, this is done by:
- checking ctx.Done() in cluster.loop and closing the cluster if ctx is cancelled
- checking if context is cancelled in OpenConn, preventing dialing after context cancellation.
- preventing topology refresh if context is cancelled, and during refresh in calls to openConnPool preventing opening new pools.

NewSession's context also stops requests from being sent to server in connWriter.send after cancellation,
contexts passed to Session.Prepare, Query.Execute, Query.AsyncExecute provide the same behaviour.
Context's cancellation is also checked right before allocating requests StreamID.

Requests that were sent but didn't reach the driver before context cancellation are met with an error response mentioning context cancellation, I'm not sure if this is right or if there should be a bit of additional time to receive the response, as the request may have been sent near context deadline leaving no room for receiving its response.

Also changed OpenConn to use DialContext and HandshakeContext

@martin-sucha @zimnx please let me know if this is the desired behavior, or if I missed something crucial

Choraden commented:
@Kulezi Can I get a little more **context** of what you did there?

Kulezi commented:
ping @martin-sucha, @zimnx

zimnx commented on experiments/cmd/benchtab/main.go:
```diff
@@ -34,7 +35,7 @@ func main() {
 
 	if !config.dontPrepare {
-		initSession, err := scylla.NewSession(cfg)
+		initSession, err := scylla.NewSession(context.Background(), cfg)
```
How about defining a context using `signal.Notify(context.Background(), ...)` and passing it everywhere instead passing uninterruptable context?

zimnx commented on session_integration_bench_test.go:
```diff
@@ -17,33 +18,33 @@ func BenchmarkSessionQueryIntegration(b *testing.B) {
 	for _, stmt := range initStmts {
 		q := session.Query(stmt)
-		if _, err := q.Exec(); err != nil {
+		if _, err := q.Exec(context.Background()); err != nil {
```
The same comment about using signal.Notify

zimnx commented on transport/cluster.go:
```diff
@@ -493,19 +494,23 @@ func (c *Cluster) handleStatusChange(v *StatusChange) {
-			c.tryReopenControl()
+			c.tryReopenControl(ctx)
+		case <-ctx.Done():
+			log.Println("cluster closing due to context cancellation")
```
libraries shouldn't log anything unless it's explicitely enabled

zimnx commented on transport/compress.go:
```diff
@@ -28,12 +30,15 @@ type compr struct {
 
 	if read <= 9 {
+		if ctx.Err() != nil {
+			return 0, fmt.Errorf("request aborted due to context cancellation")
```
context might be timed out, not cancelled so error message is wrong

zimnx commented on transport/compress.go:
```diff
@@ -28,12 +30,15 @@ type compr struct {
 	read := len(c.buf.Bytes())
 
 	if read <= 9 {
+		if ctx.Err() != nil {
```
wouldn't checking if context is done and returning wrapped error be better here?

zimnx commented on transport/compress.go:
```diff
@@ -45,6 +50,10 @@ func (c *compr) compress(dst io.Writer, src *bytes.Buffer) (written int64, err e
 		return 0, err
 	}
 
+	if ctx.Err() != nil {
```
ditto

zimnx commented on transport/conn.go:
```diff
@@ -114,11 +117,20 @@ func (c *connWriter) send(r request) error {
 	l := uint32(len(b) - frame.HeaderSize)
 	binary.BigEndian.PutUint32(b[5:9], l)
 
+	if ctx.Err() != nil {
```
ditto

zimnx commented on transport/conn.go:
```diff
@@ -171,7 +183,7 @@ func (c *connReader) handler(streamID frame.StreamID) ResponseHandler {
+func (c *connReader) loop(ctx context.Context) {
 	c.bufw = frame.BufferWriter(&c.buf)
 	for {
 		resp := c.recv()
```
ctx is unused and looking like this loop blocks on recv, which should be interruptable

Kulezi replied:
Thanks for noticing. Actually reader should not take the context as this recv can only be interrupted by its underlying connection being closed, which is already done in cluster/pool when it sees that the context is done.

zimnx commented on transport/conn.go:
```diff
@@ -639,8 +651,11 @@ func MakeResponseHandlerWithError(err error) ResponseHandler {
-func (c *Conn) sendRequest(req frame.Request, compress, tracing bool) (frame.Response, error) {
-	c.sendController()
+func (c *Conn) sendRequest(ctx context.Context, req frame.Request, compress, tracing bool) (frame.Response, error) {
+	c.sendController(ctx)
```
shouldn't `sendController` return error?

zimnx commented on transport/conn.go:
```diff
@@ -655,22 +670,29 @@ func (c *Conn) sendRequest(req frame.Request, compress, tracing bool) (frame.Res
+	case resp := <-h:
+		return resp.Response, resp.Err
+	case <-ctx.Done():
+		return nil, fmt.Errorf("no response before context cancellation")
```
context might be timed out, error message is wrong

zimnx commented on transport/conn.go:
```diff
@@ -655,22 +670,29 @@ func (c *Conn) sendRequest(req frame.Request, compress, tracing bool) (frame.Res
+		return resp.Response, resp.Err
+	case <-ctx.Done():
+		return nil, fmt.Errorf("no response before context cancellation")
+		// TODO: timeout chan?
```
TODO?

zimnx commented on transport/conn.go:
```diff
@@ -655,22 +670,29 @@ func (c *Conn) sendRequest(req frame.Request, compress, tracing bool) (frame.Res
-
+	c.sendController(ctx)
+	if ctx.Err() != nil {
+		h <- response{Err: fmt.Errorf("no response before context cancellation")}
```
ditto

zimnx commented on transport/conn.go:
```diff
@@ -695,23 +718,23 @@ control:
+func (c *Conn) sendController(ctx context.Context) {
 	for {
-		if size := c.Waiting(); size < targetWaiting {
+		if size := c.Waiting(); size < targetWaiting || ctx.Err() != nil {
 			break
 		}
 		time.Sleep(50 * time.Millisecond)
```
This should use ticker and react to context interrupt in select

zimnx commented on transport/conn_integration_test.go:
```diff
@@ -22,7 +23,7 @@ func TestOpenShardConnIntegration(t *testing.T) {
 	for i := uint16(0); i < si.NrShards; i++ {
 		si.Shard = i
-		c, err := OpenShardConn(TestHost+":19042", si, DefaultConnConfig(""))
+		c, err := OpenShardConn(context.Background(), TestHost+":19042", si, DefaultConnConfig(""))
```
ditto

zimnx commented on transport/pool.go:
```diff
@@ -23,15 +24,19 @@ type ConnPool struct {
 
-func NewConnPool(host string, cfg ConnConfig) (*ConnPool, error) {
+func NewConnPool(ctx context.Context, host string, cfg ConnConfig) (*ConnPool, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
```
no point of checking it here, you should check it just before/during/after blocking operations

zimnx commented on transport/pool.go:
```diff
@@ -192,14 +197,14 @@ func (r *PoolRefiller) onConnClose(conn *Conn) {
+func (r *PoolRefiller) loop(ctx context.Context) {
+	r.fill(ctx)
 
 	timer := time.NewTicker(fillBackoff)
```
nit: it's ticker, not timer

zimnx commented on transport/pool.go:
```diff
@@ -192,14 +197,14 @@ func (r *PoolRefiller) onConnClose(conn *Conn) {
 
 	timer := time.NewTicker(fillBackoff)
 	for {
 		select {
```
add case for ctx.Done(), you should interrupt the for loop when it happens

zimnx commented:
Please ask for review only when PR is ready to be merged. I usually don't look at the drafts at all.

Kulezi commented:
Made changes with regards to @zimnx comments, also added an integration test that checks if the behavior is correct.
Opened issue #271 regarding logs, This PR does not implement any additional wait for responses that reached the driver near contexts deadline, I suppose if it's really necessary it can be added later.

zimnx commented on transport/conn.go:
```diff
@@ -171,6 +183,7 @@ func (c *connReader) handler(streamID frame.StreamID) ResponseHandler {
 	return h
 }
 
+// loop terminates when its connection gets closed by the pool, especially when session context is done.
```
I don't see any context being used here, why there's a hidden side effect?

Kulezi replied:
The way reader loop works is that its blocking on io.Copy which from what I recall cannot be interrupted using contexts.

When session is being closed it closes all its connections, making io.Copy return an io error, closing the reader loop.
This occurs especially when session context is closed.

zimnx commented on transport/conn.go:
```diff
@@ -639,8 +652,13 @@ func MakeResponseHandlerWithError(err error) ResponseHandler {
+	if err := c.sendController(ctx); err != nil {
+		return nil, err
+	}
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
```
shouldn't sendController return error when context is cancelled? If so, this check is redundant

Kulezi replied:
This check is not redundant as sendController could finish immediately without checking the ctx,
however it could return ctx.Err() instead of nil, making this check redundant (making it easier not to forget)

Kulezi commented on transport/conn.go:
```diff
@@ -36,6 +37,8 @@ type request struct {
 	Tracing         bool
 	ResponseHandler ResponseHandler
+
+	ctx context.Context // nolint:containedctx // cancelling sending request can't be done without it.
```
Those request structs are sent to the writer loop which sends them to scylla. This is the only way of knowing if a single requests context is done inside the loop, allowing to skip sending it if the context is done.
Scenario: someone wants to cancel a single request but not the whole session.

Kulezi commented on session_integration_test.go:
```diff
@@ -403,3 +414,68 @@ func TestPrepareIntegration(t *testing.T) {
+		}
+
+		// Await schema agreement, TODO: implement true schema agreement.
+		time.Sleep(time.Second)
```
Issue is already there, https://github.com/scylladb/scylla-go-driver/issues/252, added the mention about removal.

Kulezi commented:
Applied changes suggested by @zimnx.

I also noticed that skipping a single request can increase the inflight request counter in connWriter.loop() even though it wasn't sent, fixed this by changing the behavior from returning nil to returning a custom error in this case. This way the end user will also know the reason request was skipped, and unwrap the underlying context error.

Added missing propagation of request context to compress, now it checks its error too skipping after compressing and before sending.

Added missing context.Background() to _connCloseRequest.

Choraden commented on transport/conn.go:
```diff
@@ -81,9 +96,12 @@ func (c *connWriter) loop() {
+			if err := c.send(ctx, r); err != nil {
 				r.ResponseHandler <- response{Err: fmt.Errorf("%s send: %w", c.connString(), err)}
+				if _, ok := err.(*skippedError); ok {
+					continue
```
Wouldn't it casue streamIDs leak?

Choraden replied:
ConnWriter should be able to free streamID. Some kind of a callback with pointer to connReader seems to be convenient solution, but I'm not sure if it is the best.

Kulezi replied:
Good point, this is how i'd like to resolve it, as in sendRequest we can't be sure we encounter a skippedError before looking at ctx.Done().

Kulezi commented:
Fixed the StreamID leak, and made the commit messages describe the changes in a more verbose way.

zimnx reviewed (approved):
lgtm, thanks
